store/postgres/src/entities.rs

```diff
@@ -676,21 +687,53 @@ impl Table {
         }
     }
 
-    fn count_entities(&self, conn: &PgConnection) -> Result<u64, Error> {
-        match self {
-            Table::Public(subgraph) => {
-                let count: i64 = public::entities::table
-                    .filter(public::entities::subgraph.eq(subgraph.to_string()))
-                    .count()
-                    .get_result(conn)?;
-                Ok(count as u64)
-            }
-            Table::Split(entities) => {
-                let table = entities.table.clone();
-                let count: i64 = table.count().get_result(conn)?;
-                Ok(count as u64)
-            }
-        }
+    /// Adjust the `entityCount` property of the `SubgraphDeployment` for
+    /// `subgraph` by `count`. This needs to be performed after the changes
+    /// underlying `count` have been written to the store.
+    pub(crate) fn update_entity_count(
+        &self,
+        conn: &PgConnection,
+        subgraph: SubgraphDeploymentId,
+        count: i32,
+    ) -> Result<(), StoreError> {
+        let count_query = match self {
+            Table::Public(_) => format!(
+                "select count(*) from public.entities where subgraph = '{}'",
+                &subgraph
+            ),
+            Table::Split(entities) => format!("select count(*) from {}.entities", entities.schema),
+        };
+        // The big complication in this query is how to determine what the
+        // new entityCount should be. We want to make sure that if the entityCount
+        // is NULL or the special value `00`, it gets recomputed. Using `00` here
+        // makes it possible to manually set the `entityCount` to that value
+        // to force a recount; setting it to `NULL` is not desirable since
+        // `entityCount` on the GraphQL level is not nullable, and so setting
+        // `entityCount` to `NULL` could cause errors at that layer; temporarily
+        // returning `0` is more palatable. To be exact, recounts have to be
+        // done here, from the subgraph writer.
+        //
+        // The first argument of `coalesce` will be `NULL` if the entity count
+        // is `NULL` or `00`, forcing `coalesce` to evaluate its second
+        // argument, the query to count entities. In all other cases,
+        // `coalesce` does not evaluate its second argument
+        let query = format!(
+            "
+            update subgraphs.entities
+            set data = data || (format('{{\"entityCount\":
+                                  {{ \"data\": \"%s\", 
+                                    \"type\": \"BigInt\"}}}}',
+                                  coalesce((nullif(data->'entityCount'->>'data', '00'))::numeric + $1, ({}))))::jsonb
```

Comment:
This line strikes me as too long. Can we break up the string a little?

Reply:
I tried improving it, see if you like that better

Reply:
Not super pretty but I don't have any better ideas. 😁